Marlin/SdFatUtil.cpp

```diff
@@ -33,15 +33,26 @@ int SdFatUtil::FreeRam() {
   return &top - reinterpret_cast<char*>(sbrk(0));
 }
 #else  // __arm__
-extern char *__brkval;
-extern char __bss_end;
+extern "C" {
+  extern int  __bss_end;
+  extern int* __brkval;
+};
 /** Amount of free RAM
  * \return The number of free bytes.
  */
 int SdFatUtil::FreeRam() {
-  char top;
-  return __brkval ? &top - __brkval : &top - &__bss_end;
-}
+  int free_memory;
+  if (reinterpret_cast<int>(__brkval) == 0) {
+    // if no heap use from end of bss section
+    free_memory = reinterpret_cast<int>(&free_memory)
+                  - reinterpret_cast<int>(&__bss_end);
+  } else {
+    // use from top of stack to heap
+    free_memory = reinterpret_cast<int>(&free_memory)
+                  - reinterpret_cast<int>(__brkval);
+  }
+  return free_memory;
+};
 #endif  // __arm
```

Comment:
I don't think removing things from .gitignore is a good idea. It will just
lead to unwanted files being checked into this project. What advantage is
there to removing them?

On 7 March 2015 at 00:07, Scott Lahteine <notifications@github.com> wrote:

> In Marlin/SdFatUtil.cpp
> <https://github.com/MarlinFirmware/Marlin/pull/1569#discussion_r25988774>:
>
> >  /** Amount of free RAM
> >   * \return The number of free bytes.
> >   */
> >  int SdFatUtil::FreeRam() {
> > -  char top;
> > -  return __brkval ? &top - __brkval : &top - &__bss_end;
> > -}
> > +  int free_memory;
> > +  if (reinterpret_cast<int>(__brkval) == 0) {
> > +    // if no heap use from end of bss section
> > +    free_memory = reinterpret_cast<int>(&free_memory)
> > +                  - reinterpret_cast<int>(&__bss_end);
> > +  } else {
> > +    // use from top of stack to heap
> > +    free_memory = reinterpret_cast<int>(&free_memory)
> > +                  - reinterpret_cast<int>(__brkval);
> > +  }
> > +  return free_memory;
> > +};
> >  #endif  // __arm
>
> The FreeRam code in SdFatUtil is from a newer version of the SDFat
> library, so presumably it is more up-to-date, but let me know if the code
> you are reverting here is actually better than the newer code.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/1569/files#r25988774>.
>